ssl/extensions.cc

```diff
@@ -3089,6 +3089,48 @@ bool ssl_negotiate_alps(SSL_HANDSHAKE *hs, uint8_t *out_alert,
   return true;
 }
 
+// Adds pha_ext to the client hello
+static bool ext_pha_add_clienthello(const SSL_HANDSHAKE *hs, CBB *out,
+                                    CBB *out_compressible, ssl_client_hello_type_t type)  {
+  const SSL *const ssl = hs->ssl;
+
+  // post_handshake_auth is not supported in TLS versions < 1.3. Also checking
+  // if extension has been enabled by client.
+  if(hs->min_version < TLS1_3_VERSION || ssl->s3->pha_ext != SSL_PHA_ENABLED) {
+    return true;
+  }
+
+  CBB contents;
+  if (!CBB_add_u16(out_compressible, TLSEXT_TYPE_post_handshake_auth) ||
+      !CBB_add_u16_length_prefixed(out_compressible, &contents) ||
+      !CBB_add_u16(&contents, 0 /* empty "extension_data" field */) ||
+      !CBB_flush(out_compressible)) {
+    return false;
+  }
+
+  ssl->s3->pha_ext = SSL_PHA_EXT_SENT;
+  return true;
+}
+
+// Parses pha_ext from client hello
+static bool ext_pha_parse_clienthello(SSL_HANDSHAKE *hs,
+                                      uint8_t *out_alert,
+                                      CBS *contents)  {
+  if(contents == NULL || hs->min_version < TLS1_3_VERSION) {
+    return true;
+  }
+
+  if(CBS_len(contents) == 0) {
+    return false;
+  }
```

Comment:
This seems a bit different from what OpenSSL has, is this intentional? 
https://github.com/openssl/openssl/blob/master/ssl/statem/extensions_srvr.c#L1248-L1252

Reply:
I saw CBB_add_u16_length_prefixed being used to specify the length of the contents field but in our case I guess the contents field is always 0. Updated both the parse and add functions accordingly.